src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+﻿/********************************************************
+*                                                        *
+*   © Copyright (C) Microsoft. All rights reserved.      *
+*                                                        *
+*********************************************************/
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Runtime.CompilerServices;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A <see cref="JoinableTaskFactory" />-aware semaphore that allows reentrancy without consuming another slot in the semaphore.
+    /// </summary>
+    [DebuggerDisplay(nameof(CurrentCount) + " = {" + nameof(CurrentCount) + "}")]
```

Comment:
The reason I break it out with `nameof` is in case `CurrentCount` gets renamed or removed, it ensures that `DebuggerDisplay` gets updated (either automatically, or to fix a compile error). 
I've seen too many `DebuggerDisplay` attributes with stale strings that can't evaluate at debug time and aren't caught before the code ships. I avoid it now by always using `nameof` everywhere.

Reply:
Duh.  Of course.  Carry on, nothing to see here...